### ISSUE TITLE
Proposal - Introduce IDisposable ownership annotations

### ISSUE THREAD

**Issue by jeremyVignelles** [dotnet/corefx#37873]
Hi,

I have thought about this for a while, and I finally got time for that, so I'm posting here.
Please tell me if that's not the right place to ask for that.

# Rationale

I am an average developper that wants to make sure that I don't forget to call `Dispose()` on `IDisposable`. What is great is that there's a roslyn analyzer for that : https://github.com/DotNetAnalyzers/IDisposableAnalyzers

However, sometimes, that analyzer doesn't know how things work, and a signature like

```cs
public MyDisposableClass GetInstance()
```

will trigger a warning, while sometime it's just returning a cached copy which is owned by the class.

There is no way for the analyser to know the semantic of what should be disposed and what should not.

Then came the idea that we could provide some attributes that could help the analyzer.

# The problem
How to redistribute that kind of attributes?
- Embedding the attributes with the analyzer? The analyzer itself is a development dependency, and we shouldn't leak that kind of implementation details into the project.
- Having a package dependency for this kind of annotations? What about version conflicts between two projects that uses different versions of the annotations? 
- Having a .cs file that the project that wants to check embeds, similar to LibLog ? How would the analyzer check the attributes? with only the class name?
- Maintaining external annotations like [JetBrains](https://github.com/JetBrains/ExternalAnnotations) ? That would be probably feasible, and makes it possible to annotate existing code, but we are not JetBrains...

Each of those options have drawbacks, and probably more than exposed above. Then we thought about it in a different way :
If we truly want to analyze the behavior of the code, what if we find a more universal way : introduce the annotations directly in .net.

# The proposition
Here comes the IDisposable ownership attributes. The first formal definition was made [here](https://github.com/DotNetAnalyzers/IDisposableAnalyzers/pull/130/files), and I think that the API shape is understandable.

```cs
    /// <summary>
    /// The return value must be disposed by the caller.
    /// </summary>
    [AttributeUsage(AttributeTargets.ReturnValue | AttributeTargets.Parameter, AllowMultiple = false, Inherited = true)]
    public class GivesOwnershipAttribute : Attribute
    {
    }

    /// <summary>
    /// The return value must not be disposed by the caller.
    /// </summary>
    [AttributeUsage(AttributeTargets.ReturnValue | AttributeTargets.Parameter, AllowMultiple = false, Inherited = true)]
    public class KeepsOwnershipAttribute : Attribute
    {
    }

    /// <summary>
    /// The ownership of instance is transferred and the receiver is responsible for disposing.
    /// </summary>
    [AttributeUsage(AttributeTargets.Parameter, AllowMultiple = false, Inherited = true)]
    public class TakesOwnershipAttribute : Attribute
    {
    }
```

Let's dig into this : 
By defining these attributes, we state that "The owner of the object is responsible to `Dispose()` it".

Ownership can be given (i.e. transferred to caller):

```cs
public class ResourceFactory {
   [return: GivesOwnership]
   public Resource Create()
   {
        return new MyResource();
   }
}
```

It can be taken

```cs
public class ResourceHandler : IDisposable {
   private Resource _resource;
   public void SetResource([TakesOwnership] Resource resource) => this._resource = resource;
   public void Dispose()
   {
      this._resource.Dispose();
   }
}

new ResourceHandler().SetResource(new Resource);
```

We can also say explicitely that the resource is kept:
```cs
public class ResourceCache : IDisposable {
    private Resource _resource;
    [return: KeepsOwnership]
    public Resource GetResource()
    {
       return this._resource;
    }
}
```

# Why in .NET?
Having a way to express ownership is really nice to know what is happening, and many places in the framework itself could benefit from that.

Take this code for example:

``` cs
var stream = File.Open...;
var reader = new StreamReader(stream);
```

Which one should I dispose?

If `StreamReader` declared this constructor as `TakesOwnership`, the question is no more.

Another question I often have is : "When using `IServiceCollection.AddSingleton()`, are my objects correctly `Dispose()` 'd?"

# Potential uses

- Documentation : The documentation could automatically state that the method has a special behavior regarding the object ownership
- Analyzers : IDisposableAnalyzers would be one of them, but maybe something more integrated, like in the compiler or in VS ?

What do you think?

**Comment by Gnbrkm41** [comment]
I think one use case of this could be `HttpClient` (for the aforementioned IDisposableAnalyzers), since it's one of the types that implement `IDisposable` however [it seems to be recommended to use single instance instead of instantiating a new one every time](https://aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/), hence caching.

side note regarding 'Why in .NET?' section: just dispose of both of the instances, since calling Dispose() multiple times won't cause any problems 😝 
```cs
using (Stream stream = File.Open())
using (StreamReader reader = new StreamReader(stream))
{
    // Some nice codes come here
}
```

**Comment by jeremyVignelles** [comment]
Stream is likely not the best example of usage because you can pass it into a stream reader and transfer ownership, or specify `leaveOpen`. The scenario is complex there.

If I understand what you mean, you're telling me that it doesn't hurt to dispose the item twice? Is that always true,  or only on Stream? I'm not sure that is something that is always implemented, and I'm pretty sure that I have in my code objects that are not meant to be disposed twice.

Anyway, there are still the case where you get a IDisposable object as a result of a method, and where you must not dispose it because that would otherwise close it for the rest of the code.

**Comment by Gnbrkm41** [comment]
>If I understand what you mean, you're telling me that it doesn't hurt to dispose the item twice? 

Yes, that is correct. It is advised against throwing anything in Dispose/only making it possible to disposing once to ensure cleanup of unmanaged resources. See [Implementing a Dispose method](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose?view=netframework-4.8):
>To help ensure that resources are always cleaned up appropriately, a Dispose method should be callable multiple times without throwing an exception.

So I would expect it to be always implemented in such way.
But regardless, I think it's a fair point, hence just a side note.
